src/main/java/com/nowoczesnyjunior/financialapp/expensemodule/service/ExpenseServiceImpl.java

```diff
@@ -0,0 +1,111 @@
+        this.expenseRepository = expenseRepository;
+        this.categoryRepository = categoryRepository;
+        this.userRepository = userRepository;
+        this.expenseMapper = expenseMapper;
+    }
+
+    @Override
+    public List<ExpenseDto> getExpenses(String startDate, String endDate, String categoryName, Integer topN) {
+        LocalDateTime startLocalDate = startDate != null ? getLocalDateTime(startDate) : defaultStartDate;
+        LocalDateTime endLocalDate = endDate != null ? getLocalDateTime(endDate) : defaultEndDate;
+        List<Expense> list;
+
+        if (categoryName != null) {
+            validateCategory(categoryName);
+            list = expenseRepository.findExpenseByExpenseDateBetweenAndCategory(startLocalDate, endLocalDate, categoryName);
+        } else {
+            list = expenseRepository.findExpenseByExpenseDateBetween(startLocalDate, endLocalDate);
+        }
+        return expenseMapper.expensesToDtos(list);
+    }
+
+    @Override
+    public ExpenseDto addExpense(ExpenseDto expenseDto) {
+        Expense expense = expenseMapper.dtoToModel(expenseDto);
+        // TODO: get active user from usermodule
+        User user = userRepository.findAll().get(0);
+        expense.setUser(user);
+
+        if (expenseDto.getCategory().getId() == null || !isCategoryAvailable(expenseDto.getCategory())) {
+            throw new CategoryNotFoundException();
+        }
+
+        Expense savedExpense = expenseRepository.save(expense);
+        return expenseMapper.expenseToExpenseDto(savedExpense);
+    }
+
+    @Override
+    public void deleteExpense(Long expenseId) {
+        expenseRepository.deleteById(expenseId);
+    }
+
+    @Override
+    public ExpenseDto editExpense(Long expenseId, ExpenseDto expenseDto) {
+        Optional<Expense> oldExpence = expenseRepository.findById(expenseId);
+        if (oldExpence.isPresent()) {
+            Expense expense = expenseMapper.dtoToModel(expenseDto);
+            expense.setExpenseId(expenseId);
+            expense.setUser(oldExpence.get().getUser());
+            Expense savedExpense = expenseRepository.save(expense);
+            return expenseMapper.expenseToExpenseDto(savedExpense);
+        } else {
+            throw new ObjectNotFoundException( "Expense", expenseId);
+        }
+    }
+
+    private LocalDateTime getLocalDateTime(String date){
+        try {
+            return LocalDate.parse(date).atStartOfDay();
+        } catch (Exception e) {
+            throw new InvalidDateException(date);
+        }
+    }
+
+    private void validateCategory(String categoryName) {
+        // TODO: Add UNIQUE constraint to category name
+        if (!isCategoryAvailable(categoryName)) {
+            throw new CategoryNotFoundException();
+        }
+    }
+
+
+    private boolean isCategoryAvailable(CategoryDto category) {
```

Comment:
It's also detail, but I would go with `categoryExists` or sth like that, 'Available' suggests that there might be some categories that exists, but are not available

Reply:
done